src/base/random.test.cc

```diff
@@ -0,0 +1,217 @@
+#endif
+    // Default seed is 5489
+    ASSERT_EQ(Random::globalSeed, decltype(Random::gen)::default_seed);
+    Random rng{42};
+    // Default seed is stilldefault seed 5489
+    ASSERT_EQ(Random::globalSeed, decltype(Random::gen)::default_seed);
+    // First number corresponds to seed 42
+    ASSERT_EQ(rng.random<uint64_t>(), 13930160852258120406llu);
+}
+
+/**
+ * Test that all constructed objects get
+ * added to the live instance list
+ */
+TEST(RandomConstruct, LiveInstancesAdd)
+{
+#ifdef NDEBUG
+    GTEST_SKIP() << "Skipping as assertions are "
+        "stripped out of fast builds";
+#endif
+
+    Random base_rng{};
+    Random my_rng{42};
+    Random cpy_base_rng = base_rng;
+    Random cpy_my_rng = my_rng;
+
+    // All pointers have been added
+    ASSERT_EQ(Random::instances.size(), 4);
+
+    std::set<Random*> ptrs = { &base_rng, &my_rng, &cpy_base_rng, &cpy_my_rng};
+
+    // The correct pointers have been added
+    ASSERT_EQ(std::count_if(
+        Random::instances.begin(),
+        Random::instances.end(),
+        [&](const auto & rng) { return ptrs.count(rng) != 0;}), 4);
+}
+
+/**
+ * Test that all destructed objects get
+ * removed from the live instance list
+ */
+TEST(RandomConstruct, LiveInstancesRemove)
+{
+#ifdef NDEBUG
+    GTEST_SKIP() << "Skipping as assertions are "
+        "stripped out of fast builds";
+#endif
+
+    Random * base_rng = new Random{};
+    Random * my_rng = new Random{42};
+    Random * cpy_base_rng = new Random(*base_rng);
+    Random * cpy_my_rng = new Random(*my_rng);
+    std::set<Random*> ptrs = { my_rng, cpy_base_rng, cpy_my_rng};
+
+    delete base_rng;
+    ASSERT_EQ(Random::instances.size(), 3);
+    ASSERT_EQ(std::count_if(
+        Random::instances.begin(),
+        Random::instances.end(),
+        [&](const auto & rng) { return ptrs.count(rng) != 0;}), 3);
+
+    delete my_rng;
+    ptrs.erase(my_rng);
+    ASSERT_EQ(Random::instances.size(), 2);
+    ASSERT_EQ(std::count_if(
+        Random::instances.begin(),
+        Random::instances.end(),
+        [&](const auto & rng) { return ptrs.count(rng) != 0;}), 2);
+
+    delete cpy_base_rng;
+    ptrs.erase(cpy_base_rng);
+    ASSERT_EQ(Random::instances.size(), 1);
+    ASSERT_EQ(std::count_if(
+        Random::instances.begin(),
+        Random::instances.end(),
+        [&](const auto & rng) { return ptrs.count(rng) != 0;}), 1);
+
+    delete cpy_my_rng;
+    ptrs.erase(cpy_my_rng);
+    ASSERT_EQ(Random::instances.size(), 0);
+    ASSERT_EQ(std::count_if(
+        Random::instances.begin(),
+        Random::instances.end(),
+        [&](const auto & rng) { return ptrs.count(rng) != 0;}), 0);
+}
+
+/**
+ * Test that reseeding after construction with
+ * an explicit seed works
+ */
+TEST(RandomReseed, ConstructThenReseed)
+{
+#ifdef NDEBUG
+    GTEST_SKIP() << "Skipping as assertions are "
+        "stripped out of fast builds";
+#endif
+
+    Random rng{};
+    ASSERT_EQ(rng.random<uint64_t>(), 14514284786278117030llu);
```

Comment:
Dumb question: are we sure the first value is not host dependent somehow?

Reply:
Not dumb at all : it is not !

Actually I should have known given I started the discussion here : https://github.com/orgs/gem5/discussions/1517

It appears for the test to be compiler independent (STL implementation independent), either the test should not check for the value, so check the seed and carry the seed as a debug member, or I should first propose a PR that makes calls to `Random::random<>` compiler independent.